onnxscript/rewriter/pattern_test.py

```diff
@@ -688,6 +688,39 @@ def test_model(x: FLOAT[1024], y: FLOAT[1024]) -> FLOAT[1024]:
         self.assertEqual(len(model.graph), 2)
         self.assertEqual([x.op_type for x in model.graph], ["Constant", "Identity"])
 
+    def test_or_pattern(self):
+        def source_pattern(op, x, y, bias):
+            t1 = op.MatMul(x, y)
+            t2 = op.Add(t1, bias)
+            t1_or_t2 = pattern.OrValue([t1, t2], tag_var="has_bias", tag_values=[False, True])
+            return op.Relu(t1_or_t2)
+
+        def replacement(op, x, y, bias, has_bias):
+            if has_bias:
+                return op.WithBias(x, y, bias)
+            else:
+                return op.WithoutBias(x, y)
+
+        rule = pattern.RewriteRule(source_pattern, replacement)
+
+        @script()
+        def test_model1(x: FLOAT[16, 32], y: FLOAT[32, 16]) -> FLOAT[16, 16]:
+            return op.Relu(op.MatMul(x, y))
+
+        model_proto = test_model1.to_model_proto()
+        model = ir.serde.deserialize_model(model_proto)
+        rule.apply_to_model(model)
+        self.assertEqual([x.op_type for x in model.graph], ["WithoutBias"])
+
+        @script()
+        def test_model2(x: FLOAT[16, 32], y: FLOAT[32, 16], bias: FLOAT[16]) -> FLOAT[16, 16]:
```

Comment:
Does the output of the OrValue pattern always supposed to have the same shape/type, regardless of what path is chosen? (Is this part of what you mean by unrestricted or patterns)

Reply:
No, to Shubham's question. Shape/type restrictions are al the user's responsibility. The restrictions are spelled out in the Constructor. We can do an "Or" of `Add(...)` and `Mul(...)` ... because we can distinguish between `Add` and `Mul` and choose the correct alternative. But we cannot do an OR of `Add(x, y)` and `Add(y, x)`, for example: because looking at the `Add` we cannot determine which of the two alternatives to continue the match with. So, simple summary is: the alternatives must all correspond to distinct ops.

Reply:
I think relaxing the constraints may be useful. I will try that in a separate PR. It will need refactoring the state to support backtracking (to "undo" changes done in a failed alternative).